pkg/operator/workloads/saved_base_workload.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2019 Cortex Labs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workloads
+
+import (
+	"github.com/cortexlabs/cortex/pkg/lib/aws"
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	"github.com/cortexlabs/cortex/pkg/lib/parallel"
+	"github.com/cortexlabs/cortex/pkg/operator/config"
+	ocontext "github.com/cortexlabs/cortex/pkg/operator/context"
+)
+
+func uploadBaseWorkload(baseWorkload *BaseWorkload) error {
+	if isBaseWorkloadCached(baseWorkload) {
+		return nil
+	}
+
+	key := ocontext.BaseWorkloadKey(baseWorkload.WorkloadID, baseWorkload.AppName)
+	err := config.AWS.UploadJSONToS3(baseWorkload, key)
+	if err != nil {
+		return errors.Wrap(err, "upload base workload", baseWorkload.AppName, baseWorkload.WorkloadID)
+	}
+	cacheBaseWorkload(baseWorkload)
+	return nil
+}
+
+func uploadBaseWorkloads(baseWorkloads []*BaseWorkload) error {
+	fns := make([]func() error, len(baseWorkloads))
+	for i, baseWorkload := range baseWorkloads {
+		fns[i] = uploadBaseWorkloadFunc(baseWorkload)
+	}
+	return parallel.RunFirstErr(fns...)
+}
+
+func uploadBaseWorkloadsFromWorkloads(workloads []Workload) error {
+	fns := make([]func() error, len(workloads))
+	for i, workload := range workloads {
+		fns[i] = uploadBaseWorkloadFunc(workload.GetBaseWorkloadPtr())
+	}
+	return parallel.RunFirstErr(fns...)
+}
+
+func uploadBaseWorkloadFunc(baseWorkload *BaseWorkload) func() error {
```

Comment:
i think it may be clearer to not have this, so its even more obvious in the caller functions that these are functions, your call though

Reply:
I agree, I actually used to have it that way, however there is a nasty bug when you in-line it: On line 44, by passing in `baseWorkload` to this function, it makes a copy of the pointer. Otherwise, when `baseWorkload` iterates in the for loop, it will be updated as it goes, and when you run `parallel.RunFirstErr()`, all of the functions will run on the last one.

Here's a demo: https://play.golang.org/p/FbqIyqlCIID

It's also similar to this fun one: https://play.golang.org/p/y81_Zh4jwMi